petab/simulate.py

```diff
@@ -0,0 +1,154 @@
+import abc
+import numpy as np
+import pandas as pd
+import petab
+import shutil
+import sympy as sp
+import tempfile
+from typing import Dict, Optional
+
+
+class Simulator(abc.ABC):
+    """
+    Base class that specific simulators should inherit.
+    Specific simulators should minimally implement the
+    `_simulate_without_noise` method.
+    TODO link to amici.petab_simulate.PEtabSimulation as example
+    """
+    def __init__(self,
+                 petab_problem: petab.Problem,
+                 working_dir: Optional[str] = None):
+        """
+        Initializes the simulator with sufficient information to perform
+        a simulation. If no working directory is specified, a temporary one is
+        created.
+
+        Arguments:
+            petab_problem:
+                A PEtab problem.
+            working_dir:
+                All simulator-specific output files will be saved here.
+        """
+        self.petab_problem = petab_problem
+        if working_dir is None:
+            working_dir = tempfile.mkdtemp()
+        self.working_dir = working_dir
+        self.noise_formulas = petab.calculate.get_symbolic_noise_formulas(
+                self.petab_problem.observable_df)
+        self.rng = np.random.default_rng()
+
+    def clean_working_dir(self):
+        """
+        Deletes simulator-specific output files and their parent folder (see
+        the `__init__` method arguments).
+        """
+        shutil.rmtree(self.working_dir)
```

Comment:
This looks reeeeeally dangerous. Should at least be clearly documented in the constructor that this may be "cleaned" later on. Potentially issue a warning if it already exists / is non-empty?

Reply:
Thanks, good point, new logic is to delete the directory if it was created as a temporary directory, else only delete if `force=True`.